Add Shub Image Caching

michaelmoore10 opened the pull request:
**Description of the Pull Request (PR):**

Image downloads using the Shub client (SingularityHub and sregistry) do not take advantage of the local cache, leading to repeated downloads of the same image.  This PR adds the logic to check for the existence of the image in the cache and uses it if it matches.

**This fixes or addresses the following GitHub issues:**

- Fixes #2697


**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](../CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](../CHANGELOG.md) if necessary according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](../CONTRIBUTORS.md)


Attn: @singularity-maintainers

michaelmoore10 committed:
Add logic to check the local cache for SingularityHub/Sregistry downloads

michaelmoore10 committed:
Added name to contributors list

bauerm97 commented on cmd/internal/cli/actions.go:
```diff
@@ -154,7 +154,16 @@ func handleShub(u string) (string, error) {
 	imagePath := cache.ShubImage("hash", imageName)
 
-	libexec.PullShubImage(imagePath, u, true, noHTTPS)
+    exists, err := cache.ShubImageExists("hash", imageName)
```
Hey @michaelmoore10 , the code looks good to me. Can you make sure you run `go fmt ./...` with the `go1.11` version of `go` installed and then push that as a new commit? Your code looks to be incorrectly formatted, and `go` is very particular about formatting. Cheers, looking forward to reviewing once that's fixed :)

michaelmoore10 replied:
Hi @bauerm97,

  Thanks!  I pushed a new version of this fix with the correct formatting.  

-Mike

michaelmoore10 committed:
Fixed formatting

bauerm97 commented on cmd/internal/cli/actions.go:
```diff
@@ -154,7 +154,16 @@ func handleShub(u string) (string, error) {
+		return "", fmt.Errorf("unable to check if %v exists: %v", imagePath, err)
+	}
+	if !exists {
+		sylog.Infof("Downloading shub image")
```
I think this should be `Verbosef`.

bauerm97 commented on cmd/internal/cli/actions.go:
```diff
@@ -154,7 +154,16 @@ func handleShub(u string) (string, error) {
+		sylog.Infof("Downloading shub image")
+		libexec.PullShubImage(imagePath, u, true, noHTTPS)
+	} else {
+		sylog.Infof("Use image from cache")
```
Same comment here, probably should be `Verbosef`. Does a user expect these to print when running without any logging level commands? What's your opinion here?

michaelmoore10 replied:
For both of these changes, the behavior is consistent with the handleLibrary() and handleNet() routines.  I agree, at least for the "Use image from cache" log message that Verbosef() is better.  No need to have that spit out for normal use.

Should the other routines be changed as well for using the image in the cache? And for the "downloading image" message in all of the routines? Is it ok to change the existing default behavior for logging?

bauerm97 replied:
>I agree, at least for the "Use image from cache" log message that Verbosef() is better. No need to have that spit out for normal use.

Why don't we go ahead and make that change then. I think it's not so useful to write that out

>Should the other routines be changed as well for using the image in the cache? And for the "downloading image" message in all of the routines? Is it ok to change the existing default behavior for logging?

I think it's okay if we change that. Maybe we will first merge this Pull Request as is, then you can follow up with a second pull request changing all of the logging messages? Are you up for that @michaelmoore10 ?

michaelmoore10 replied:
Yep.  I'm up for that.

michaelmoore10 committed:
In handleShub(), changed Infof() log message to Verbosef() when the
requested image is in the cache. Reduces the standard logging messages.
This same change will be applied later to the other handle...() cases
in a future PR to make the behavior consistent

bauerm97 reviewed (approved):
Looks good to me then!

bauerm97 committed:
Merge branch 'master' into fix/shub-caching
